src/plugins/sqlite.js

```diff
@@ -0,0 +1,126 @@
+"use strict";
+
+const path = require("path");
+const fsextra = require("fs-extra");
+const sqlite3 = require("sqlite3");
+const Helper = require("../helper");
+const Msg = require("../models/msg");
+
+const currentSchemaVersion = 1;
+
+const schema = [
+	// Schema version #1
+	"CREATE TABLE IF NOT EXISTS options (name TEXT, value TEXT, CONSTRAINT name_unique UNIQUE (name))",
+	"CREATE TABLE IF NOT EXISTS messages (network TEXT, channel TEXT, time INTEGER, type TEXT, msg TEXT)",
+	"CREATE INDEX IF NOT EXISTS network_channel ON messages (network, channel)",
+	"CREATE INDEX IF NOT EXISTS time ON messages (time)",
+];
+
+class MessageStorage {
+	constructor() {
+		this.isEnabled = false;
+	}
+
+	enable(name) {
+		const logsPath = path.join(Helper.getHomePath(), "logs");
+		const sqlitePath = path.join(logsPath, `${name}.sqlite3`);
+
+		try {
+			fsextra.ensureDirSync(logsPath);
+		} catch (e) {
+			log.error("Unable to create logs directory", e);
+
+			return;
+		}
+
+		this.isEnabled = true;
+
+		this.database = new sqlite3.cached.Database(sqlitePath);
+		this.database.serialize(() => {
+			schema.forEach((line) => this.database.run(line));
+
+			this.database.get("SELECT value FROM options WHERE name = 'schema_version'", (err, row) => {
+				if (err) {
+					return log.error(`Failed to retrieve schema version: ${err}`);
+				}
+
+				// New table
+				if (row === undefined) {
+					this.database.serialize(() => this.database.run("INSERT INTO options (name, value) VALUES ('schema_version', ?)", currentSchemaVersion));
+
+					return;
+				}
+
+				const storedSchemaVersion = parseInt(row.value, 10);
+
+				if (storedSchemaVersion === currentSchemaVersion) {
+					return;
+				}
+
+				if (storedSchemaVersion > currentSchemaVersion) {
+					return log.error(`sqlite messages schema version is higher than expected (${storedSchemaVersion} > ${currentSchemaVersion}). Is The Lounge out of date?`);
+				}
+
+				log.info(`sqlite messages schema version is out of date (${storedSchemaVersion} < ${currentSchemaVersion}). Running migrations if any.`);
+
+				this.database.serialize(() => this.database.run("UPDATE options SET value = ? WHERE name = 'schema_version'", currentSchemaVersion));
+			});
+		});
+	}
+
+	index(network, channel, msg) {
+		if (!this.isEnabled) {
+			return;
+		}
+
+		const clonedMsg = Object.keys(msg).reduce((newMsg, prop) => {
+			// id is regenerated when messages are retrieved
+			// previews are not stored because storage is cleared on lounge restart
```

Comment:
Do we have a way to re-trigger previews when they are loaded? (i.e. if user never loads older messages, they wouldn't get re-fetched)

Reply:
We technically can, but I wouldn't look at it because it's not as simple as iterating all links and fetching previews (would need to queue and process with limits).

Reply:
Makes sense, I didn't mean in this PR necessarily.